python/hail/methods/statgen.py

```diff
@@ -2186,28 +2184,94 @@ def balding_nichols_model(n_populations, n_samples, n_variants, n_partitions=Non
-                                            af_dist._jrep(),
-                                            seed,
-                                            reference_genome._jrep,
-                                            mixture)
-    return MatrixTable(jmt)
-
-@typecheck(mt=MatrixTable,f=anytype)
+        fst = [0.1 for _ in range(n_populations)]
+
+    if n_partitions is None:
+        n_partitions = max(8, int(n_samples * n_variants / 1000000))
+
+    # verify args
+    for name, var in {"populations": n_populations,
+                      "samples": n_samples,
+                      "variants": n_variants,
+                      "partitions": n_partitions}.items():
+        if var < 1:
+            raise ValueError("n_{} must be positive, got {}".format(name, var))
+
+    for name, var in {"pop_dist": pop_dist, "fst": fst}.items():
+        if len(var) != n_populations:
+            raise ValueError("{} must be of length n_populations={}, got length {}"
+                             .format(name, n_populations, len(var)))
+
+    if any(map(lambda x: x < 0, pop_dist)):
+        raise ValueError("pop_dist must be non-negative, got {}"
+                         .format(pop_dist))
+
+    if any(map(lambda x: x <= 0 or x >= 1, fst)):
+        raise ValueError("elements of fst must satisfy 0 < x < 1, got {}"
+                         .format(fst))
+
+    # verify af_dist and seed
+    if not af_dist._is_constant:
+        raise ExpressionException('balding_nichols_model expects af_dist to ' +
+                                  'have constant arguments: found expression ' +
+                                  'from source {}'
+                                  .format(af_dist._indices.source))
+
+    if not isinstance(af_dist._ir, hl.ir.ApplySeeded) or af_dist.dtype != tfloat64:
+        raise ValueError("af_dist must be a random function with return type tfloat64.")
+
+    fn = af_dist._ir.function
+    import inspect
+    params = eval('list(inspect.signature(hl.{}).parameters.keys())'.format(fn))
+    args = list(map(lambda arg: construct_expr(arg, tfloat64).value, af_dist._ir.args))
+    assert params[-1] == 'seed'
+    params = params[:-1]
+    struct_af_dist = hl.struct(type=fn, **{p: a for p, a in zip(params, args)}, seed=af_dist._ir.seed)
+
+    info("balding_nichols_model: generating genotypes for {} populations, {} samples, and {} variants..."
+         .format(n_populations, n_samples, n_variants))
+
+    # generate matrix table
+    rand = hl.utils.HailSeedGenerator(seed)
+
+    bn = hl.utils.range_matrix_table(n_variants, n_samples, n_partitions)
+    bn = bn.annotate_globals(n_populations=n_populations,
+                             n_samples=n_samples,
+                             n_variants=n_variants,
+                             n_partitions=n_partitions,
+                             pop_dist=pop_dist,
+                             fst=fst,
+                             ancestral_af_dist=struct_af_dist,
```

Comment:
I think we just drop this, the user controls the `af_dist`, it's not our responsibility to include that information in the globals. They can record it, if they want, in whatever idiosyncratic way they like.